### PR TITLE
Better serializer registration, get more than just the first module

### PR DESCRIPTION
**Pull request by bf4**
But is potentially breaking anyone on rc3, but the fix is just
to manually register the adapter with the rc3-style name

### CODE REVIEW

File: lib/active_model/serializer/adapter.rb [hunk 1]
```diff
@@ -77,7 +81,7 @@ def find_by_name(adapter_name)
 
       # Automatically register adapters when subclassing
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/adapter.rb]: that's pretty cool.

File: lib/active_model/serializer/adapter.rb [hunk 2]
```diff
@@ -41,8 +41,12 @@ def adapters
         # @param [Class] klass - adapter class itself
         # @example
         #     AMS::Adapter.register(:my_adapter, MyAdapter)
+        # @note The registered name strips out 'ActiveModel::Serializer::Adapter::'
+        #   so that registering 'ActiveModel::Serializer::Adapter::Json' and
+        #   'Json' will both register as 'json'.
         def register(name, klass)
-          adapter_map.update(name.to_s.underscore => klass)
+          name = name.to_s.remove('ActiveModel::Serializer::Adapter::'.freeze)
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/adapter.rb]: why freeze the string? it's not gonna be used again? or does ruby cache it for the next time register is called? (cause that would be cool)
Also, if that's the case, should the string move to a class constant? like `ADAPTER_PREFIX` or something?

File: test/action_controller/serialization_test.rb [hunk 3]
```diff
@@ -79,8 +79,15 @@ def render_object_expired_with_cache_enabled
...
           post.title = 'ZOMG a New Post'
-          sleep 0.1
-          render json: post
+
+          expires_in = [
+            PostSerializer._cache_options[:expires_in],
+            CommentSerializer._cache_options[:expires_in],
+          ].max + 200
+
+          Timecop.travel(Time.zone.now + expires_in) do
```
Comment (NullVoxPopuli) [inline test/action_controller/serialization_test.rb]: Timecop or Delorean or good gems for time tests :-)

File: test/action_controller/serialization_test.rb [hunk 4]
```diff
@@ -307,7 +314,13 @@ def test_render_with_cache_enable_and_expired
         }
 
         assert_equal 'application/json', @response.content_type
-        assert_equal expected.to_json, @response.body
+        actual   = @response.body
+        expected = expected.to_json
+        if ENV['APPVEYOR'] && actual != expected
+          skip('Cache expiration tests sometimes fail on Appveyor. FIXME :)')
```
Comment (NullVoxPopuli) [inline test/action_controller/serialization_test.rb]: :-(

File: test/serializers/adapter_for_test.rb [hunk 5]
```diff
@@ -138,14 +138,14 @@ def test_inherited_adapter_hooks_register_adapter
...
       end
 
-      def test_inherited_adapter_hooks_register_demodulized_adapter
+      def test_inherited_adapter_hooks_register_namespaced_adapter
         Object.const_set(:MyNamespace, Module.new)
         MyNamespace.const_set(:MyAdapter, Class.new)
         my_adapter = MyNamespace::MyAdapter
         ActiveModel::Serializer::Adapter.inherited(my_adapter)
-        assert_equal ActiveModel::Serializer::Adapter.lookup(:my_adapter), my_adapter
+        assert_equal ActiveModel::Serializer::Adapter.lookup(:'my_namespace/my_adapter'), my_adapter
```
Comment (NullVoxPopuli) [inline test/serializers/adapter_for_test.rb]: Nice, I think this resolves an issue someone brought up earlier in the week User::SomethingSerializer

File: lib/active_model/serializer/adapter.rb (hunk 2 above)
Comment (bf4) [inline lib/active_model/serializer/adapter.rb]: I don't think it's worth a constant, but yes, in later Rubies frozen strings reduce object allocations

File: test/action_controller/serialization_test.rb (hunk 3 above)
Comment (bf4) [inline test/action_controller/serialization_test.rb]: Not sure I understand this comment

File: lib/active_model/serializer/adapter.rb (hunk 2 above)
Comment (NullVoxPopuli) [inline lib/active_model/serializer/adapter.rb]: legit

File: test/action_controller/serialization_test.rb (hunk 3 above)
Comment (NullVoxPopuli) [inline test/action_controller/serialization_test.rb]: I mistyped. Just saying I like the use of the Timecop or Delorean gems when it comes to test things that are time sensitive.

**Comment by NullVoxPopuli** [comment]
looks like there was an error with a bunch of the tests.

File: test/action_controller/serialization_test.rb (hunk 4 above)
Comment (bf4) [inline test/action_controller/serialization_test.rb]: weird, that shouldn't be in here

**Comment by NullVoxPopuli** [comment]
all on rails 4.0

**Comment by bf4** [comment]
history got borked somehwere

**Comment by NullVoxPopuli** [comment]
:-(

**Comment by bf4** [comment]
fixed

**Comment by bf4** [comment]
Weird appveyor fails on SimpleCov 58% on Rails 4.0.. no idea why

**Comment by NullVoxPopuli** [comment]
Travis is still failingmon rails 4.0 as well.

Can you run against 4.0 locally and see if that'll give you any insight? 

*crosses fingers*

**Comment by bf4** [comment]
@NullVoxPopuli Since it's different, I usually develop with `export RAILS_VERSION=4.0`, looks like I forgot

**Comment by NullVoxPopuli** [comment]
Thanks for fixing!

**Pull request merged**